# **Product Data Cleaning By Hephzibah Reginald Otuene**

## Task:
**Data Cleaning, Preparation, and Title Optimization**

## Objective:
Prepare raw marketing data for analysis by addressing data quality issues and creating a new feature, short_title, for improved SEO and readability. The task involves resolving data quality issues like missing values and duplicates, standardizing data formats, and generating concise product titles for better marketing impact.

## Dataset:
Product Data ([Dataset Link](https://docs.google.com/spreadsheets/d/1z3Io4xtC2FMif2lskLbDDcRe5OTjoi4l/edit?rtpof=true&sd=true))

## Task Overview
This task focuses on ensuring the dataset is clean, reliable, and ready for further marketing analysis while introducing a new short_title feature for SEO-optimized and concise product titles. You will explore, identify, and resolve common data issues using tools like Python (Pandas, NumPy) or Excel. Additionally, you'll implement logic to generate shorter product titles for each entry, ensuring they retain key information while being concise and prepare a detailed technical report.

### Steps to Complete the Task
1. Dataset Familiarization
- Download and review the dataset to understand its structure and key variables.
- Identify significant variables for cleaning and for creating the short_title feature.
- Check column names for clarity and consistency.
2. Initial Data Exploration
- Conduct a high-level review of the dataset using descriptive statistics.
- Identify data quality issues such as missing values, duplicates, and inconsistent formats.
- Use Python (Pandas, NumPy) or Excel for initial exploration.
3. Data Cleaning Process
- Handling Missing Values: Identify and evaluate columns with missing data.
- Apply appropriate strategies.
- Removing Duplicate Entries: Locate and remove duplicate records to ensure data accuracy.
- Standardize column names (e.g., sales_amount instead of SalesAmount).
- Verifying Data Accuracy: Check critical variables for anomalies (e.g., negative prices or invalid ratings).
4. Creating the short_title Feature
- Objective: Generate a concise version of product titles that retains essential information for SEO and readability.
- Steps: Analyze original product titles to identify key components (e.g., product name, category, attributes).
- Remove redundant phrases or words (e.g., "includes," "set of," "features").
- Limit titles to 30–50 characters, focusing on essential details and keywords.
- Implementation Example:
- Original Title: "Tulip Flowers Blackout Curtain for Door, Window & Room | Eyelets & Tie Back | Canvas Fabric | Set of 2 PCS"
- Short Title: "Tulip Blackout Curtain - 2 PCS"
- Original Title: "Marks & Spencer Girls' Pyjama Sets T86_2561C_Navy Mix_9-10Y"
- Short Title: "Girls' Navy Pyjama Set - 9-10Y"
5. Documentation and Reporting
- Prepare a professional report documenting the data cleaning and title optimization process:
- Introduction: Briefly describe the dataset and task objectives.
- Data Cleaning: Summarize issues identified and the cleaning steps taken.
- Short Title Creation: Explain the methodology and examples of optimized titles.
- Clean Dataset Overview: Highlight key statistics and improvements in the dataset.
- Include screenshots or visualizations illustrating the impact of data cleaning and title optimization.
6. Review and Submission
- Proofread and format the report for clarity and professionalism.
- Submit the cleaned dataset, including the short_title column, and the report.


## Let's Begin

In [63]:
import pandas as pd
import numpy as np
import re

# Load dataset
df = pd.read_excel("productdata.xlsx")

# Display first few rows
df.head()


PRODUCTID                                              TITLE  \
0    1925202  ArtzFolio Tulip Flowers Blackout Curtain for D...   
1    2673191  Marks & Spencer Girls' Pyjama Sets T86_2561C_N...   
2    2765088  PRIKNIK Horn Red Electric Air Horn Compressor ...   
3    1594019  ALISHAH Women's Cotton Ankle Length Leggings C...   
4     283658  The United Empire Loyalists: A Chronicle of th...   

                                       BULLET_POINTS  \
0  [LUXURIOUS & APPEALING: Beautiful custom-made ...   
1  [Harry Potter Hedwig Pyjamas (6-16 Yrs),100% c...   
2  [Loud Dual Tone Trumpet Horn, Compatible With ...   
3  [Made By 95%cotton and 5% Lycra which gives yo...   
4                                                NaN   

                                         DESCRIPTION  PRODUCTTYPEID  \
0                                                NaN         1650.0   
1                                                NaN         2755.0   
2  Specifications: Color: Red, Material: Aluminiu...         7537.0   
3  AISHAH Women's Lycra Cotton Ankel Leggings. Br...         2996.0   
4                                                NaN         6112.0   

   ProductLength  
0    2125.980000  
1     393.700000  
2     748.031495  
3     787.401574  
4     598.424000

In [64]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3847 entries, 0 to 3846
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PRODUCTID      3847 non-null   int64  
 1   TITLE          3847 non-null   object 
 2   BULLET_POINTS  2256 non-null   object 
 3   DESCRIPTION    1703 non-null   object 
 4   PRODUCTTYPEID  3669 non-null   float64
 5   ProductLength  3669 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 180.5+ KB


In [65]:
# Check column names
df.columns

# Check for missing values
df.isnull().sum()

# Check for duplicates and print them
duplicates = df[df.duplicated(subset=['PRODUCTID'], keep=False)]
print("Duplicate Entries:")
print(duplicates)

# Count duplicates
df.duplicated(subset=['PRODUCTID']).sum()


Duplicate Entries:
      PRODUCTID                                              TITLE  \
12       648364  J'ecris des lettres! premiers exercices d'écr...   
13      1991694  Mediterranean diet for beginners: 7Benefits of...   
14      2790448  SEGOVIA Single Walled Stainless Steel Sports| ...   
15      1810976  Stone & Beam Fan Embossed Planter in Blue, Medium   
16      1262926    Star Trek 50th Anniversary Cereamic Storage Jar   
...         ...                                                ...   
3842      67078  Occupational Health and Safety: International ...   
3843     653247                                Le duel introuvable   
3844    1783479  ArtToFrames 27x39 Inch Black Picture Frame, Th...   
3845     770603  The Ascension of Isaiah: translated from the E...   
3846     112393                Crossings: Selected Travel Writings   

                                          BULLET_POINTS  \
12                                                  NaN   
13                    

306

### This dataset has 306 duplicates using the PRODUCTID

In [66]:
# Standardize column names
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

# Rename specific columns
df.rename(columns={
    'productid': 'product_id',
    'bullet_points': 'bullet_points',
    'description': 'description',
    'producttypeid': 'product_type_id',
    'productlength': 'product_length'
}, inplace=True)

# Drop duplicates
df.drop_duplicates(subset=['product_id'], keep='first', inplace=True)

# Drop rows where title is missing
df.dropna(subset=['title'], inplace=True)

#save the data set
df.to_excel('processed_dataset.xlsx', index=False)

# Confirm cleaning
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 3541 entries, 0 to 3562
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   product_id       3541 non-null   int64  
 1   title            3541 non-null   object 
 2   bullet_points    2178 non-null   object 
 3   description      1645 non-null   object 
 4   product_type_id  3541 non-null   float64
 5   product_length   3541 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 193.6+ KB



### **306 rows were deleted and the new file saved**

Next we work on missing values. 
From the info above;
1. **bullet_points** has **2178** non-null values, meaning **1363** rows need filling.
2. **description** has **1645** non-null values, meaning **1896** rows need filling.
   
we'd do the filling using the title column which had no missing values

In [67]:
import pandas as pd

# Load dataset
dg = pd.read_excel("processed_dataset.xlsx")

# Function to generate bullet points from the title
def generate_bullet_points(title):
    if pd.isna(title):
        return "No bullet points available"
    
    keywords = title.split()[:5]  # Extract first 5 words
    return [
        f"Premium quality {keywords[0]} product",
        f"Designed for {keywords[1]} users" if len(keywords) > 1 else "High performance design",
        f"Suitable for {keywords[2]} applications" if len(keywords) > 2 else "Multi-purpose usage",
        f"Crafted with durable materials for long-lasting use",
        f"Available in multiple sizes and colors"
    ]

# Function to generate a meaningful description
def generate_description(title):
    if pd.isna(title):
        return "No description available"
    return f"'{title}' is a high-quality product designed for durability, performance, and customer satisfaction."

# Fill missing bullet_points
dg.loc[dg['bullet_points'].isna(), 'bullet_points'] = dg.loc[dg['bullet_points'].isna(), 'title'].apply(
    lambda x: ', '.join(generate_bullet_points(x))
)

# Fill missing descriptions
dg.loc[dg['description'].isna(), 'description'] = dg.loc[dg['description'].isna(), 'title'].apply(generate_description)

# Display the first few rows to verify
dg.to_excel('filled_dataset.xlsx', index=False)
dg.head()


product_id                                              title  \
0     1925202  ArtzFolio Tulip Flowers Blackout Curtain for D...   
1     2673191  Marks & Spencer Girls' Pyjama Sets T86_2561C_N...   
2     2765088  PRIKNIK Horn Red Electric Air Horn Compressor ...   
3     1594019  ALISHAH Women's Cotton Ankle Length Leggings C...   
4      283658  The United Empire Loyalists: A Chronicle of th...   

                                       bullet_points  \
0  [LUXURIOUS & APPEALING: Beautiful custom-made ...   
1  [Harry Potter Hedwig Pyjamas (6-16 Yrs),100% c...   
2  [Loud Dual Tone Trumpet Horn, Compatible With ...   
3  [Made By 95%cotton and 5% Lycra which gives yo...   
4  Premium quality The product, Designed for Unit...   

                                         description  product_type_id  \
0  'ArtzFolio Tulip Flowers Blackout Curtain for ...             1650   
1  'Marks & Spencer Girls' Pyjama Sets T86_2561C_...             2755   
2  Specifications: Color: Red, Material: Aluminiu...             7537   
3  AISHAH Women's Lycra Cotton Ankel Leggings. Br...             2996   
4  'The United Empire Loyalists: A Chronicle of t...             6112   

   product_length  
0     2125.980000  
1      393.700000  
2      748.031495  
3      787.401574  
4      598.424000

In [68]:
dg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3541 entries, 0 to 3540
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   product_id       3541 non-null   int64  
 1   title            3541 non-null   object 
 2   bullet_points    3541 non-null   object 
 3   description      3541 non-null   object 
 4   product_type_id  3541 non-null   int64  
 5   product_length   3541 non-null   float64
dtypes: float64(1), int64(2), object(3)
memory usage: 166.1+ KB


### Now, the missing columns have been filled.

### *Let's* work on the next step

## Creating the short_title column

In [69]:
import re
import pandas as pd

# Load dataset
dh = pd.read_excel("filled_dataset.xlsx")

# Function to generate a meaningful short title
def generate_short_title(title):
    if pd.isna(title):
        return "No short title available"
    
    # Step 1: Remove unwanted details (dimensions, numbers, special characters)
    title_cleaned = re.sub(r'\b\d+(\.\d+)?(cm|inch|feet|pcs|set|ml|l|kg|g|Y)?\b', '', title, flags=re.IGNORECASE)
    
    # Step 2: Remove extra spaces and symbols
    title_cleaned = re.sub(r'[_|]', ' ', title_cleaned)  # Replace underscores/pipes with spaces
    title_cleaned = re.sub(r'\s+', ' ', title_cleaned).strip()

    # Step 3: Extract meaningful words
    words = title_cleaned.split()

    # Define stop words that don't add value to short titles
    stop_words = {"for", "with", "and", "-", "by", "the", "of", "to", "on", "in"}

    # Define product keywords to ensure we stop at the right place
    product_keywords = {"curtain", "pyjama", "horn", "shirt", "dress", "pot", "stand", "bucket", "laptop", "case", "bag", "toy", "shoes"}

    # Step 4: Extract first meaningful words, stopping at product keyword
    short_title_words = []
    for word in words:
        if word.lower() not in stop_words:
            short_title_words.append(word)
        if word.lower() in product_keywords:  # Stop when reaching a product keyword
            break
        if len(short_title_words) >= 5:  # Max words before stopping
            break

    # Step 5: Append quantity/set info if available
    match = re.search(r'(\d+\s?(pcs|set|pack|Y))', title, re.IGNORECASE)
    if match:
        short_title_words.append(f"- {match.group(1)}")  # Append formatted quantity/size info

    return ' '.join(short_title_words)

# Apply function to create 'short_title' column
dh['short_title'] = dh['title'].apply(generate_short_title)

# Save dataset with new column
dh.to_excel('short_title_dataset.xlsx', index=False)

# Display first few rows
dh.head()


product_id                                              title  \
0     1925202  ArtzFolio Tulip Flowers Blackout Curtain for D...   
1     2673191  Marks & Spencer Girls' Pyjama Sets T86_2561C_N...   
2     2765088  PRIKNIK Horn Red Electric Air Horn Compressor ...   
3     1594019  ALISHAH Women's Cotton Ankle Length Leggings C...   
4      283658  The United Empire Loyalists: A Chronicle of th...   

                                       bullet_points  \
0  [LUXURIOUS & APPEALING: Beautiful custom-made ...   
1  [Harry Potter Hedwig Pyjamas (6-16 Yrs),100% c...   
2  [Loud Dual Tone Trumpet Horn, Compatible With ...   
3  [Made By 95%cotton and 5% Lycra which gives yo...   
4  Premium quality The product, Designed for Unit...   

                                         description  product_type_id  \
0  'ArtzFolio Tulip Flowers Blackout Curtain for ...             1650   
1  'Marks & Spencer Girls' Pyjama Sets T86_2561C_...             2755   
2  Specifications: Color: Red, Material: Aluminiu...             7537   
3  AISHAH Women's Lycra Cotton Ankel Leggings. Br...             2996   
4  'The United Empire Loyalists: A Chronicle of t...             6112   

   product_length                                       short_title  
0     2125.980000  ArtzFolio Tulip Flowers Blackout Curtain - 2 PCS  
1      393.700000               Marks & Spencer Girls' Pyjama - 10Y  
2      748.031495                                      PRIKNIK Horn  
3      787.401574               ALISHAH Women's Cotton Ankle Length  
4      598.424000              United Empire Loyalists: A Chronicle

In [70]:
dh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3541 entries, 0 to 3540
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   product_id       3541 non-null   int64  
 1   title            3541 non-null   object 
 2   bullet_points    3541 non-null   object 
 3   description      3541 non-null   object 
 4   product_type_id  3541 non-null   int64  
 5   product_length   3541 non-null   float64
 6   short_title      3541 non-null   object 
dtypes: float64(1), int64(2), object(4)
memory usage: 193.8+ KB


In [71]:
# Save cleaned dataset
dh.to_excel("cleaned_dfa.xlsx", index=False)

print("Data cleaning completed! File saved as 'cleaned_dfa.xlsx'")


Data cleaning completed! File saved as 'cleaned_dfa.xlsx'


## Thank you; i'm open to corrections and suggestions.
*Do have a nice day*